fmf/base.py

```diff
@@ -584,16 +589,22 @@ def apply_rules(rule_set):
                 # Apply remaining rule attributes if context matches
                 try:
                     if context.matches(condition):
-                        if decision_callback:
-                            decision_callback(self, rule, True)
-
                         # Remove special keys (when, because...) from the rule
                         apply_rule = {
                             key: value
                             for key, value in rule.items()
                             if key not in ADJUST_CONTROL_KEYS
                             }
-                        self._merge_special(self.data, apply_rule)
+
+                        if decision_callback:
+                            node_before = self.copy()
```

Comment:
Should it `deep_copy` instead in case the merge changes something like a list, or is that already taken care of?